BringYourOwnBoard.md

```diff
@@ -0,0 +1,126 @@
+
+# On Patches In this Repo
+
+There are a lot of patches within this repository. This is a normal process of developing meta-layers for Yocto. They're intended only for the `meta-raspberrypi` layer. You don't need to think about them, however you might find that you do need to update and or edit your existing BSP with some patches of your own. You can safely disregard them in your port. 
+
+# Recipes
+
+## recipes-bsp 
+
+### What does it do? 
+
+Within our project `recipes-bsp` is used for appending instructions to the bootloader that allow SwUpdate to set which partition to boot into after completing an image update (eg if we've completed an update while running on `/dev/mmcblk0p0` then when the device restarts we boot into `/dev/mmcblk0p1` which contains the new version of the image). For our purposes we can use [rpi-u-boot-scr.bbappend](./recipes-bsp/rpi-u-boot-scr/rpi-u-boot-scr.bbappend) to add the [boot.cmd.in](./recipes-bsp/rpi-u-boot-scr/files/boot.cmd.in) file to the Raspberry Pi 4's boot process which instructs the device to do the A/B switching. 
+
+Within this layer we also edit u-boot to use our own firmware environment configuration and append a special patch for a gcc issue that RaspberryPi has been dealing with. 
+### What do I need to think about? 
+
+You must talk to your BSP provider and determine what bootloader they provide as apart of the BSP. Likely you will need to create a more durable u-boot script that handles detecting if an update occurred, determines what partition to load into, and provide the logic to do that switching. 
+
+For instance you can look at this snippet: 
+```u-boot
+fdt addr ${fdt_addr} && fdt get value bootargs /chosen bootargs
+fatload mmc 0:1 ${kernel_addr_r} @@KERNEL_IMAGETYPE@@
+if env exists rpipart;then setenv bootargs ${bootargs} root=/dev/mmcblk0p${rpipart}; fi
+if test ! -e mmc 0:1 uboot.env; then saveenv; fi;
+@@KERNEL_BOOTCMD@@ ${kernel_addr_r} - ${fdt_addr}
+```
+
+Within the [recipes-core] layer we create three partitions and mount them in the `fstab` : `/dev/mmcblk0p2` (the 'A' partition), `/dev/mmcblk0p3` (the 'B' partition), and `/dev/mmcblk0p4` (the "adu" partition). When an update occurs we write the partition number to boot into next to the u-boot environment section setup using `fw-env.config` which is then used by this u-boot script to determine what partition to load into on the next boot. 
+
+For a production device the core idea of using a flag to say what to load into is completely acceptable, however there are some safety checks you should add if you want your device to be secure and durable. 
+
+First you should also write the hash of the image to be booted into. Then u-boot should hash the partition you're attempting to write into (at least include a CRC checksum) and determine if the partition is what you expect. This prevents direct flash injection attacks and keeps your device from running unsafe code. 
```

Comment:
For this: _Then u-boot should hash the partition you're attempting to write into (at least include a CRC checksum) and determine if the partition is what you expect. This prevents direct flash injection attacks and keeps your device from running unsafe code._ I'm curious if another reason is to detect any corruption before you write something to the partition that won't work? Or is the hashing only useful as a security check?

Reply:
This is a bootloader step. SwUpdate handles checking the hash of the *.swu file that contains the image that has been sent down to the device during the update step. 

If the file was good at import it is guaranteed to be good at install (if the update is successful. we fail the update if the contents coming down through the pipe do not match the hash in the update manifest or the hash in the *.swu file). 

You are right. In aerospace applications where you have to worry about radiation and EMI you usually do this before run time to make sure you didn't get a fried transistor/ NAND flash at some point, however in most terrestrial  applications we're more concerned about a bad actor with a scalpel and a magnet trying to inject code into the bootloader to compromise the device (turn off device, disasseble the bootloader, find load location, load your own OS at that location, and now you have access to all the data on the device). 

The real way to prevent these kinds of attacks is to have a protected chip that stores the hashes of the A and the B sides. It also usually stores the bootloader itself (because you can corrupt that too if the bad actor finds a JTAG port that hasn't been turned off). When the device boots this separate chip will read the flash of the partition it is supposed to boot into and then check if it's secured. 

I actually think this is how parts of secure boot work in many BIOSs. 

That reminds me. We should mention that delta update extends the life of flash based devices. 

Swupdate delta is pretty easy to work with. I've been reading more of their documentation and our own. It'd be interesting to look at what a better default integration would look like.